deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,683 @@
+		context->error = status;
+		context->error_message = strerror(status);
+		return STATUS_FAILURE;
+	}
+	return status;
+}
+
+int happy_eyeballs_timedwait_default(struct happy_eyeballs_ctx *context)
+{
+	return happy_eyeballs_timedwait(context,
+					HAPPY_EYEBALLS_CONNECTION_TIMEOUT_MS);
+}
+
+int happy_eyeballs_timedwait(struct happy_eyeballs_ctx *context,
+			     unsigned long time_in_millis)
+{
+	if (context == NULL)
+		return STATUS_INVALID_ARGUMENT;
+
+	int status = os_event_timedwait(context->race_completed_event,
+					time_in_millis);
+
+	if (context->error != 0)
+		return STATUS_FAILURE;
+
+	if (status != 0 && status != ETIMEDOUT) {
+		context->error = status;
+		return STATUS_FAILURE;
+	}
+	return status;
+}
+
+int happy_eyeballs_destroy(struct happy_eyeballs_ctx *context)
+{
+	if (context == NULL)
+		return STATUS_INVALID_ARGUMENT;
+
+#ifdef _WIN32
+#define SHUT_RDWR SD_BOTH
+#else
+#define closesocket(s) close(s)
+#endif
+	// We do not need to lock context->candidate_mutex here because the candidate array is _only_ mutated
+	// in happy_eyeballs_connect. Using this function at the same time as connect is a programmer error.
+	// Locking it here will cause a deadlock with join.
+
+	// Shutdown non-winning sockets (but keep the socket alive so that things error out in threads)
+	for (int i = 0; i < context->candidates.num; i++) {
+		if (context->candidates.array[i].sockfd != INVALID_SOCKET &&
+		    context->candidates.array[i].sockfd != context->socket_fd) {
+			shutdown(context->candidates.array[i].sockfd,
+				 SHUT_RDWR);
+		}
+	}
+	// Join threads
+	for (int i = 0; i < context->candidates.num; i++) {
+		pthread_join(context->candidates.array[i].thread, NULL);
+		os_event_destroy(
+			context->candidates.array[i].socket_completed_event);
+	}
+	// Close sockets
+	for (int i = 0; i < context->candidates.num; i++) {
+		if (context->candidates.array[i].sockfd != INVALID_SOCKET &&
+		    context->candidates.array[i].sockfd != context->socket_fd) {
+			closesocket(context->candidates.array[i].sockfd);
+		}
+	}
+	pthread_mutex_destroy(&context->winner_mutex);
+	pthread_mutex_destroy(&context->candidate_mutex);
+	os_event_destroy(context->race_completed_event);
+	if (context->addresses != NULL)
+		freeaddrinfo(context->addresses);
+
+	da_free(context->candidates);
+	free(context);
+	return STATUS_SUCCESS;
+}
+
+/**
+ * Setters & Getters
+ */
+
+int happy_eyeballs_set_bind_addr(struct happy_eyeballs_ctx *context,
+				 socklen_t addr_len,
+				 struct sockaddr_storage *addr_storage)
+{
+	if (!context)
+		return STATUS_INVALID_ARGUMENT;
+
+	if (addr_storage && addr_len > 0) {
+		memcpy(&context->bind_addr, addr_storage,
+		       sizeof(struct sockaddr_storage));
+		context->bind_addr_len = addr_len;
+	} else {
+		context->bind_addr_len = 0;
+		memset(&context->bind_addr, 0, sizeof(struct sockaddr_storage));
+	}
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_get_socket_fd(struct happy_eyeballs_ctx *context)
```

Comment:
```suggestion
int happy_eyeballs_get_socket_fd(const struct happy_eyeballs_ctx *context)
```
Shouldn't make much of a difference, but might as well mark this as `const` since no mutation is happening